grafana/resource_organization.go

```diff
@@ -0,0 +1,304 @@
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	d.Set("name", resp.Name)
+	if err := ReadUsers(d, meta); err != nil {
+		return err
+	}
+	return nil
+}
+
+func UpdateOrganization(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+	orgId, _ := strconv.ParseInt(d.Id(), 10, 64)
+	if d.HasChange("name") {
+		name := d.Get("name").(string)
+		err := client.UpdateOrg(orgId, name)
+		if err != nil {
+			return err
+		}
+	}
+	return UpdateUsers(d, meta)
+}
+
+func DeleteOrganization(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+	orgId, _ := strconv.ParseInt(d.Id(), 10, 64)
+	return client.DeleteOrg(orgId)
+}
+
+func ExistsOrganization(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*gapi.Client)
+	orgId, _ := strconv.ParseInt(d.Id(), 10, 64)
+	_, err := client.Org(orgId)
+	if err != nil && err.Error() == "404 Not Found" {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, err
+}
+
+func ImportOrganization(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	exists, err := ExistsOrganization(d, meta)
+	if err != nil || !exists {
+		return nil, errors.New("Error Importing Grafana Organization")
+	}
+	d.Set("admin_user", "admin")
+	d.Set("create_users", "true")
+	err = ReadOrganization(d, meta)
+	if err != nil {
+		return nil, err
+	}
+	return []*schema.ResourceData{d}, nil
+}
+
+func ReadUsers(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+	orgId, _ := strconv.ParseInt(d.Id(), 10, 64)
+	orgUsers, err := client.OrgUsers(orgId)
+	if err != nil {
+		return err
+	}
+	roleMap := map[string][]string{"Admin": nil, "Editor": nil, "Viewer": nil}
+	grafAdmin := d.Get("admin_user")
+	for _, orgUser := range orgUsers {
+		if orgUser.Login != grafAdmin {
+			roleMap[orgUser.Role] = append(roleMap[orgUser.Role], orgUser.Email)
+		}
+	}
+	for k, v := range roleMap {
+		d.Set(fmt.Sprintf("%ss", strings.ToLower(k)), v)
+	}
+	return nil
+}
+
+func UpdateUsers(d *schema.ResourceData, meta interface{}) error {
+	oldUsers, newUsers := collectUsers(d)
+	changes := changes(oldUsers, newUsers)
+	orgId, _ := strconv.ParseInt(d.Id(), 10, 64)
+	changes, err := addIds(d, meta, changes)
+	if err != nil {
+		return err
+	}
+	return applyChanges(meta, orgId, changes)
+}
+
+func collectUsers(d *schema.ResourceData) (map[string]OrgUser, map[string]OrgUser) {
+	roles := []string{"admins", "editors", "viewers"}
+	oldUsers, newUsers := make(map[string]OrgUser), make(map[string]OrgUser)
+	for _, role := range roles {
+		roleName := strings.Title(role[:len(role)-1])
+		old, new := d.GetChange(role)
+		for _, u := range old.([]interface{}) {
+			oldUsers[u.(string)] = OrgUser{0, u.(string), roleName}
+		}
+		for _, u := range new.([]interface{}) {
+			newUsers[u.(string)] = OrgUser{0, u.(string), roleName}
```

Comment:
I hadn't considered that- yes it does, and then flip-flops on each apply... I'll add a sanity check to prevent an email being specified multiple times within an org

Reply:
I also changed the `old`/`new` terminology to `state`/`config` to clarify that we're comparing between Grafana's state and the desired configuration